### ISSUE TITLE
Add images

### ISSUE THREAD

**Issue by OpenMatt** [MozillaFoundation/mofo-lightsaber#5]
plug-ins that could help:

```
ep_copy_paste_images Copy and Paste images from clipboard to a pad

ep_embedmedia Embed media (youtube, vimeo etc)
```

**Comment by OpenMatt** [comment]
This one is huge -- people would really love to be able to embed images. We got embedded videos working, but couldn't get images to work yet.

**Comment by JohnMcLear** [comment]
I did a LOT of work around images, a few weeks..

Images are a fsking nightmare in editors, especially those that are contentEditable.

I can implement basic image functionality but here are the things that users expect that are no longer being developed in contentEditable..
- Image Resizing: works in FF and IE, not Chrome.  Catching events is a nightmare.  UX is horrible in all browsers.
- Image Drag/Drop: Works in FF afaik but creates a duplicate, probably solvable.
- Image alignment: Impossibru in Etherpad due to it's per line nature.

Trying to put a shim/solution in place that wraps handles around an image is a nightmare too as you end up with a lot of horrible cruft in the DOM.  

There needs to be some consideration on if the image should be uploaded and stored on the file system too or if the base64 encoded content of the image should be saved as an changeset ergo in the HTML.. Saving Base64 would make export a lot easier (no external assets) but has it's own caveats...

Basically I can implement this as a plugin and maybe only allow image modification functionality in the browsers that support it.

It's at least a week dedicated work though..

**Comment by jdotpz** [comment]
ep_embedmedia is still behaving as we saw last time, with images being hosed but being able to embed other types of media.

**Comment by JohnMcLear** [comment]
I'm working on Image support now.  See the copy paste images plugin.  Support welcome
----- Reply message -----

**Comment by jdotpz** [comment]
![screen shot 2015-01-28 at 10 31 29 am](https://cloud.githubusercontent.com/assets/1373628/5941798/d9e15bec-a6d8-11e4-8f34-1f88cf25eea4.JPG)
I got that as I tried to install the copy paste image plugin.  Awesome that you are cooking on that!

**Comment by JohnMcLear** [comment]
You will need latest develop and it's a moving target.  I only have one more days funding to finish the to-do so let's hope things go well later today ..
----- Reply message -----

**Comment by JohnMcLear** [comment]
You can test image support here: https://beta.etherpad.org/p/imagedemo

**Comment by jdotpz** [comment]
Nice, it's looking great!

**Comment by OpenMatt** [comment]
@jdotpz should we mark this one closed?

**Comment by OpenMatt** [comment]
- From testing today, images sometimes appear in doubles or multiples. e.g., you paste one image, and it appears on the pad 3 times.

**Comment by JohnMcLear** [comment]
https://github.com/JohnMcLear/ep_copy_paste_images is the repo

Afaik I fixed copy/paste in FF

Please update Etherpad and the plugin, tnx :)

**Comment by JohnMcLear** [comment]
Yea it looks again like you are potentially running an old version, testing on https://beta.etherpad.org/p/image seems to show this issue resolved.

**Comment by OpenMatt** [comment]
- EtherpadDotOrg also commented on our test pad and twitter to say that there is an update that fixes this

**Comment by OpenMatt** [comment]
Noticing issues like this one:
https://dl.dropboxusercontent.com/spa/6c38yp3crbxni5b/wk__vcww.png

**Comment by JohnMcLear** [comment]
@OpenMatt wow how did you make it do that?  That looks like a browser bug, I'm running nightly here and don't see that.

**Comment by OpenMatt** [comment]
@JohnMcLear yeah that was probably a browser bug
- I've been working with images in the pad quite a bit the last few days, and it feels stable. 
- @jdotpz I think maybe we can mark this one resolved, unless you have more stuff you want to do here?

**Comment by jdotpz** [comment]
+1
